File: packages/persona-core/src/persona_core.mjs
/**
 * persona-core -- reference implementation
 *
 * Implements PersonaPack helpers and in-memory registry primitives described in:
 * `packages/persona-core/product_spec.md`
 * `packages/persona-core/tech_spec.md`
 */

import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";

const PERSONA_PACK_SCHEMA_URL = new URL("../schemas/persona_pack.schema.json", import.meta.url);

export const CAMERA_MODES = Object.freeze(["A_SELFIE", "B_MIRROR", "C_CUTAWAY"]);

export const DEFAULT_ANCHOR_REFRESH_POLICY = Object.freeze({
  refresh_every_turns: 8,
  drift_fail_threshold: 0.74,
  drift_warn_threshold: 0.84,
  flicker_fail_threshold: 0.6,
});

/**
 * @typedef {Object} PersonaPack
 * @property {string} persona_id
 * @property {string} version
 * @property {Record<string, AnchorEntry[]>} anchor_sets
 * @property {Object} identity
 * @property {string[]=} identity.face_embedding_refs
 * @property {Object} style
 * @property {Object=} voice_profile
 * @property {Object} behavior_policy
 */

/**
 * @typedef {Object} AnchorEntry
 * @property {string} image_ref
 * @property {string=} mask_ref
 * @property {Object} metadata
 * @property {string=} metadata.expression_tag
 * @property {Object=} metadata.head_pose
 * @property {string=} metadata.lighting_tag
 * @property {number[]=} metadata.crop_box
 * @property {string[]=} metadata.best_for
 */

/**
 * @typedef {Object} DriftSignal
 * @property {number=} identity_similarity
 * @property {number=} bg_similarity
 * @property {number=} flicker_score
 */

/**
 * @typedef {Object} ActorTimelineEvent
 * @property {number=} t0
 * @property {number=} t1
 * @property {string=} emotion
 * @property {number=} intensity
 * @property {string=} gaze_mode
 * @property {number=} blink_rate
 * @property {Object=} head_motion
 * @property {"listening"|"speaking"=} state
 */

/**
 * @returns {string}
 */
export function readPersonaPackSchemaText() {
  return readFileSync(fileURLToPath(PERSONA_PACK_SCHEMA_URL), "utf8");
}

/**
 * @returns {any}
 */
export function readPersonaPackSchemaJson() {
  return JSON.parse(readPersonaPackSchemaText());
}

/**
 * Lightweight PersonaPack validation (shape-level).
 * @param {any} pack
 * @returns {{ ok: boolean, errors: string[] }}
 */
export function validatePersonaPack(pack) {
  /** @type {string[]} */
  const errors = [];
  if (pack === null || typeof pack !== "object") errors.push("pack must be an object");
  if (!pack?.persona_id || typeof pack.persona_id !== "string") errors.push("persona_id must be a string");
  if (!pack?.version || typeof pack.version !== "string") errors.push("version must be a string");
  if (!pack?.anchor_sets || typeof pack.anchor_sets !== "object") errors.push("anchor_sets must be an object");
  if (!pack?.identity || typeof pack.identity !== "object") errors.push("identity must be an object");
  if (!pack?.style || typeof pack.style !== "object") errors.push("style must be an object");
  if (!pack?.behavior_policy || typeof pack.behavior_policy !== "object") errors.push("behavior_policy must be an object");

  if (pack?.anchor_sets && typeof pack.anchor_sets === "object") {
    for (const [mode, anchors] of Object.entries(pack.anchor_sets)) {
      if (!Array.isArray(anchors) || anchors.length === 0) errors.push(`anchor_sets.${mode} must be a non-empty array`);
      for (let i = 0; i < (anchors ?? []).length; i++) {
        const anchor = anchors[i];
        if (!anchor?.image_ref) errors.push(`anchor_sets.${mode}[${i}].image_ref is required`);
      }
    }
  }

  return { ok: errors.length === 0, errors };
}

/**
 * @param {PersonaPack} pack
 * @param {string} mode
 * @returns {AnchorEntry[]}
 */
export function getAnchorSet(pack, mode) {
  if (pack?.anchor_sets?.[mode]) return pack.anchor_sets[mode];
  const first = Object.values(pack?.anchor_sets ?? {})[0];
  return Array.isArray(first) ? first : [];
}

/**
 * @param {AnchorEntry[]} anchors
 * @returns {AnchorEntry | null}
 */
export function selectCanonicalAnchor(anchors) {
  if (!Array.isArray(anchors) || anchors.length === 0) return null;
  const canonical = anchors.find((a) => a.metadata?.best_for?.includes("canonical"));
  if (canonical) return canonical;
  const fallback = anchors.find((a) => a.metadata?.best_for?.includes("default"));
  return fallback ?? anchors[0];
}

/**
 * @param {AnchorEntry} anchor
 * @param {string=} desiredEmotion
 * @returns {number}
 */
export function scoreAnchor(anchor, desiredEmotion) {
  let score = 0;
  if (desiredEmotion && anchor.metadata?.expression_tag) {
    if (anchor.metadata.expression_tag.toLowerCase() === desiredEmotion.toLowerCase()) score += 2;
  }
  if (desiredEmotion && anchor.metadata?.best_for?.some((tag) => tag.toLowerCase() === desiredEmotion.toLowerCase())) {
    score += 1;
  }
  if (anchor.metadata?.best_for?.includes("canonical")) score += 0.25;
  return score;
}

/**
 * @param {{
 *  drift?: DriftSignal,
 *  turn_index?: number,
 *  policy?: Partial<typeof DEFAULT_ANCHOR_REFRESH_POLICY>
 * }} args
 * @returns {{ refresh: boolean, reason: string }}
 */
export function shouldRefreshAnchor(args) {
  const policy = { ...DEFAULT_ANCHOR_REFRESH_POLICY, ...(args.policy ?? {}) };
  const drift = args.drift ?? {};

  if (typeof drift.identity_similarity === "number" && drift.identity_similarity < policy.drift_fail_threshold) {
    return { refresh: true, reason: "identity_fail" };
  }
  if (typeof drift.bg_similarity === "number" && drift.bg_similarity < policy.drift_fail_threshold) {
    return { refresh: true, reason: "background_fail" };
  }
  if (typeof drift.flicker_score === "number" && drift.flicker_score > policy.flicker_fail_threshold) {
    return { refresh: true, reason: "flicker_fail" };
  }
  const turnIndex = args.turn_index ?? 0;
  if (policy.refresh_every_turns > 0 && turnIndex > 0 && turnIndex % policy.refresh_every_turns === 0) {
    return { refresh: true, reason: "periodic_refresh" };
  }
  return { refresh: false, reason: "stable" };
}

/**
 * Select an anchor image for the given mode with continuity + drift refresh support.
 *
 * @param {{
 *  persona_pack: PersonaPack,
 *  mode: string,
 *  desired_emotion?: string,
 *  last_anchor_ref?: string,
 *  drift?: DriftSignal,
 *  turn_index?: number,
 *  policy?: Partial<typeof DEFAULT_ANCHOR_REFRESH_POLICY>
 * }} args
 * @returns {{ anchor: AnchorEntry | null, mode: string, reason: string }}
 */
export function selectAnchor(args) {
  const anchors = getAnchorSet(args.persona_pack, args.mode);
  if (!anchors.length) return { anchor: null, mode: args.mode, reason: "no_anchors" };

  const refresh = shouldRefreshAnchor({ drift: args.drift, turn_index: args.turn_index, policy: args.policy });
  if (!refresh.refresh && args.last_anchor_ref) {
    const last = anchors.find((a) => a.image_ref === args.last_anchor_ref);
    if (last) return { anchor: last, mode: args.mode, reason: "reuse_last_anchor" };
  }

  if (refresh.refresh) {
    const canonical = selectCanonicalAnchor(anchors);
    return { anchor: canonical, mode: args.mode, reason: `refresh:${refresh.reason}` };
  }

  const sorted = [...anchors].sort((a, b) => {
    const scoreDiff = scoreAnchor(b, args.desired_emotion) - scoreAnchor(a, args.desired_emotion);
    if (scoreDiff !== 0) return scoreDiff;
    return String(a.image_ref).localeCompare(String(b.image_ref));
  });

  return { anchor: sorted[0], mode: args.mode, reason: "best_match" };
}

/**
 * Clamp actor timeline against persona behavior policy.
 * @param {ActorTimelineEvent[]} timeline
 * @param {{
 *  emotion_ranges?: Record<string, { min?: number, max?: number }>,
 *  allowed_emotions?: string[]
 * }} behaviorPolicy
 * @returns {ActorTimelineEvent[]}
 */
export function clampActorTimeline(timeline, behaviorPolicy) {
  const allowed = behaviorPolicy?.allowed_emotions?.map((e) => String(e)) ?? null;
  const ranges = behaviorPolicy?.emotion_ranges ?? {};

  return (timeline ?? []).map((event) => {
    const next = { ...event };
    if (allowed && next.emotion && !allowed.includes(next.emotion)) next.emotion = allowed[0] ?? "neutral";

    const range = next.emotion ? ranges[next.emotion] : null;
    if (range && typeof next.intensity === "number") {
      const min = range.min ?? 0;
      const max = range.max ?? 1;
      next.intensity = Math.max(min, Math.min(max, next.intensity));
    } else if (typeof next.intensity === "number") {
      next.intensity = Math.max(0, Math.min(1, next.intensity));
    }
    return next;
  });
}

/**
 * In-memory Persona registry for testing and local prototyping.
 */
export class PersonaRegistry {
  /** @param {{ asset_resolver?: (ref: string) => string }=} opts */
  constructor(opts) {
    this.asset_resolver = opts?.asset_resolver ?? null;
    /** @type {Map<string, { metadata?: any, versions: Map<string, PersonaPack> }>} */
    this.personas = new Map();
  }

  /** @param {{ persona_id: string, metadata?: any }} args */
  createPersona(args) {
    if (!args?.persona_id) throw new Error("persona_id is required");
    if (this.personas.has(args.persona_id)) throw new Error(`persona already exists: ${args.persona_id}`);
    this.personas.set(args.persona_id, { metadata: args.metadata ?? {}, versions: new Map() });
  }

  /** @param {{ persona_id: string, pack: PersonaPack }} args */
  createPersonaVersion(args) {
    if (!args?.persona_id) throw new Error("persona_id is required");
    const entry = this.personas.get(args.persona_id);
    if (!entry) throw new Error(`persona not found: ${args.persona_id}`);

    const validation = validatePersonaPack(args.pack);
    if (!validation.ok) throw new Error(`invalid PersonaPack: ${validation.errors.join("; ")}`);
    if (args.pack.persona_id !== args.persona_id) throw new Error("persona_id mismatch in PersonaPack");

    entry.versions.set(args.pack.version, args.pack);
    return args.pack.version;
  }

  /** @param {{ persona_id: string, version: string }} args */
  getPersonaPack(args) {
    const entry = this.personas.get(args.persona_id);
    if (!entry) return null;
    return entry.versions.get(args.version) ?? null;
  }

  /** @param {string} personaId */
  listPersonaVersions(personaId) {
    const entry = this.personas.get(personaId);
    if (!entry) return [];
    return Array.from(entry.versions.keys());
  }

  /** @param {string} ref */
  resolveAsset(ref) {
    if (!this.asset_resolver) return ref;
    return this.asset_resolver(ref);
  }
}

File: packages/persona-core/python/persona_core.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Callable, Dict, List, Optional


CAMERA_MODES = ["A_SELFIE", "B_MIRROR", "C_CUTAWAY"]

DEFAULT_ANCHOR_REFRESH_POLICY = {
    "refresh_every_turns": 8,
    "drift_fail_threshold": 0.74,
    "drift_warn_threshold": 0.84,
    "flicker_fail_threshold": 0.6,
}


def _schema_path() -> Path:
    return Path(__file__).resolve().parents[1] / "schemas" / "persona_pack.schema.json"


def read_persona_pack_schema_text() -> str:
    return _schema_path().read_text("utf-8")


def read_persona_pack_schema_json() -> Dict[str, Any]:
    return json.loads(read_persona_pack_schema_text())


def validate_persona_pack(pack: Any) -> Dict[str, Any]:
    errors: List[str] = []
    if pack is None or not isinstance(pack, dict):
        errors.append("pack must be an object")
        return {"ok": False, "errors": errors}

    if not isinstance(pack.get("persona_id"), str) or not pack.get("persona_id"):
        errors.append("persona_id must be a string")
    if not isinstance(pack.get("version"), str) or not pack.get("version"):
        errors.append("version must be a string")
    if not isinstance(pack.get("anchor_sets"), dict):
        errors.append("anchor_sets must be an object")
    if not isinstance(pack.get("identity"), dict):
        errors.append("identity must be an object")
    if not isinstance(pack.get("style"), dict):
        errors.append("style must be an object")
    if not isinstance(pack.get("behavior_policy"), dict):
        errors.append("behavior_policy must be an object")

    anchor_sets = pack.get("anchor_sets")
    if isinstance(anchor_sets, dict):
        for mode, anchors in anchor_sets.items():
            if not isinstance(anchors, list) or len(anchors) == 0:
                errors.append(f"anchor_sets.{mode} must be a non-empty array")
                continue
            for idx, anchor in enumerate(anchors):
                if not isinstance(anchor, dict) or not anchor.get("image_ref"):
                    errors.append(f"anchor_sets.{mode}[{idx}].image_ref is required")

    return {"ok": len(errors) == 0, "errors": errors}


def get_anchor_set(pack: Dict[str, Any], mode: str) -> List[Dict[str, Any]]:
    anchor_sets = pack.get("anchor_sets") if isinstance(pack, dict) else None
    if isinstance(anchor_sets, dict) and mode in anchor_sets:
        anchors = anchor_sets.get(mode)
        return anchors if isinstance(anchors, list) else []
    if isinstance(anchor_sets, dict):
        for anchors in anchor_sets.values():
            if isinstance(anchors, list):
                return anchors
    return []


def select_canonical_anchor(anchors: List[Dict[str, Any]]) -> Optional[Dict[str, Any]]:
    if not anchors:
        return None
    canonical = next(
        (anchor for anchor in anchors if "canonical" in (anchor.get("metadata", {}).get("best_for") or [])),
        None,
    )
    if canonical:
        return canonical
    fallback = next(
        (anchor for anchor in anchors if "default" in (anchor.get("metadata", {}).get("best_for") or [])),
        None,
    )
    return fallback or anchors[0]


def score_anchor(anchor: Dict[str, Any], desired_emotion: Optional[str]) -> float:
    score = 0.0
    if desired_emotion:
        desired = desired_emotion.lower()
        expression = anchor.get("metadata", {}).get("expression_tag")
        if isinstance(expression, str) and expression.lower() == desired:
            score += 2.0
        best_for = anchor.get("metadata", {}).get("best_for") or []
        if any(isinstance(tag, str) and tag.lower() == desired for tag in best_for):
            score += 1.0
    best_for = anchor.get("metadata", {}).get("best_for") or []
    if "canonical" in best_for:
        score += 0.25
    return score


def should_refresh_anchor(args: Dict[str, Any]) -> Dict[str, Any]:
    policy = dict(DEFAULT_ANCHOR_REFRESH_POLICY)
    policy.update(args.get("policy") or {})
    drift = args.get("drift") or {}

    identity_similarity = drift.get("identity_similarity")
    if isinstance(identity_similarity, (int, float)) and identity_similarity < policy["drift_fail_threshold"]:
        return {"refresh": True, "reason": "identity_fail"}
    bg_similarity = drift.get("bg_similarity")
    if isinstance(bg_similarity, (int, float)) and bg_similarity < policy["drift_fail_threshold"]:
        return {"refresh": True, "reason": "background_fail"}
    flicker_score = drift.get("flicker_score")
    if isinstance(flicker_score, (int, float)) and flicker_score > policy["flicker_fail_threshold"]:
        return {"refresh": True, "reason": "flicker_fail"}

    turn_index = int(args.get("turn_index") or 0)
    refresh_every = int(policy.get("refresh_every_turns", 0) or 0)
    if refresh_every > 0 and turn_index > 0 and turn_index % refresh_every == 0:
        return {"refresh": True, "reason": "periodic_refresh"}

    return {"refresh": False, "reason": "stable"}


def select_anchor(args: Dict[str, Any]) -> Dict[str, Any]:
    anchors = get_anchor_set(args.get("persona_pack") or {}, args.get("mode") or "")
    if not anchors:
        return {"anchor": None, "mode": args.get("mode"), "reason": "no_anchors"}

    refresh = should_refresh_anchor(
        {
            "drift": args.get("drift"),
            "turn_index": args.get("turn_index"),
            "policy": args.get("policy"),
        }
    )
    last_anchor_ref = args.get("last_anchor_ref")
    if not refresh.get("refresh") and last_anchor_ref:
        last = next((anchor for anchor in anchors if anchor.get("image_ref") == last_anchor_ref), None)
        if last:
            return {"anchor": last, "mode": args.get("mode"), "reason": "reuse_last_anchor"}

    if refresh.get("refresh"):
        canonical = select_canonical_anchor(anchors)
        return {
            "anchor": canonical,
            "mode": args.get("mode"),
            "reason": f"refresh:{refresh.get('reason')}",
        }

    sorted_anchors = sorted(
        anchors,
        key=lambda anchor: (
            -score_anchor(anchor, args.get("desired_emotion")),
            str(anchor.get("image_ref") or ""),
        ),
    )
    return {"anchor": sorted_anchors[0], "mode": args.get("mode"), "reason": "best_match"}


def clamp_actor_timeline(
    timeline: List[Dict[str, Any]],
    behavior_policy: Dict[str, Any],
) -> List[Dict[str, Any]]:
    allowed = behavior_policy.get("allowed_emotions")
    allowed_emotions = [str(item) for item in allowed] if isinstance(allowed, list) else None
    ranges = behavior_policy.get("emotion_ranges") or {}

    clamped: List[Dict[str, Any]] = []
    for event in timeline or []:
        next_event = dict(event)
        emotion = next_event.get("emotion")
        if allowed_emotions and emotion and emotion not in allowed_emotions:
            next_event["emotion"] = allowed_emotions[0] if allowed_emotions else "neutral"

        emotion_key = next_event.get("emotion")
        if emotion_key in ranges and isinstance(next_event.get("intensity"), (int, float)):
            bounds = ranges.get(emotion_key) or {}
            min_intensity = bounds.get("min", 0)
            max_intensity = bounds.get("max", 1)
            next_event["intensity"] = max(min_intensity, min(max_intensity, next_event["intensity"]))
        elif isinstance(next_event.get("intensity"), (int, float)):
            next_event["intensity"] = max(0, min(1, next_event["intensity"]))

        clamped.append(next_event)
    return clamped


@dataclass
class PersonaRegistryEntry:
    metadata: Dict[str, Any]
    versions: Dict[str, Dict[str, Any]]


class PersonaRegistry:
    def __init__(self, asset_resolver: Optional[Callable[[str], str]] = None) -> None:
        self.asset_resolver = asset_resolver
        self.personas: Dict[str, PersonaRegistryEntry] = {}

    def create_persona(self, persona_id: str, metadata: Optional[Dict[str, Any]] = None) -> None:
        if not persona_id:
            raise ValueError("persona_id is required")
        if persona_id in self.personas:
            raise ValueError(f"persona already exists: {persona_id}")
        self.personas[persona_id] = PersonaRegistryEntry(metadata=metadata or {}, versions={})

    def create_persona_version(self, persona_id: str, pack: Dict[str, Any]) -> str:
        if not persona_id:
            raise ValueError("persona_id is required")
        entry = self.personas.get(persona_id)
        if not entry:
            raise ValueError(f"persona not found: {persona_id}")

        validation = validate_persona_pack(pack)
        if not validation["ok"]:
            raise ValueError(f"invalid PersonaPack: {'; '.join(validation['errors'])}")
        if pack.get("persona_id") != persona_id:
            raise ValueError("persona_id mismatch in PersonaPack")

        version = pack.get("version")
        if not isinstance(version, str) or not version:
            raise ValueError("version must be a string")

        entry.versions[version] = pack
        return version

    def get_persona_pack(self, persona_id: str, version: str) -> Optional[Dict[str, Any]]:
        entry = self.personas.get(persona_id)
        if not entry:
            return None
        return entry.versions.get(version)

    def list_persona_versions(self, persona_id: str) -> List[str]:
        entry = self.personas.get(persona_id)
        if not entry:
            return []
        return list(entry.versions.keys())

    def resolve_asset(self, ref: str) -> str:
        if not self.asset_resolver:
            return ref
        return self.asset_resolver(ref)
